third_party/airflow/armada/operators/armada.py

```diff
@@ -0,0 +1,97 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from airflow.models import BaseOperator
+from airflow.exceptions import AirflowException
+
+from armada_client.client import ArmadaClient
+from armada.operators.jobservice import JobServiceClient
+
+import logging
+
+from armada.operators.utils import airflow_error, search_for_job_complete
+
+armada_logger = logging.getLogger("airflow.task")
+
+
+class ArmadaOperator(BaseOperator):
+    """
+    Implementation of an ArmadaOperator for airflow.
+
+    Airflow operators inherit from BaseOperator.
+
+    :param name: The name of the airflow task
+    :param armada_client: The Armada Python GRPC client
+                        that is used for interacting with Armada
+    :param job_service_client: The JobServiceClient that is used for polling
+    :param queue: The queue name
+    :param job_set_id: The job_set_id. Should be set at dag level for all jobs
+    :param job_request_items: A PodSpec that is used by Armada for submitting a job
+
+    :return: a job service client instance
+    """
+
+    def __init__(
+        self,
+        name: str,
+        armada_client: ArmadaClient,
+        job_service_client: JobServiceClient,
+        queue: str,
+        job_set_id: str,
+        job_request_items,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.name = name
+        self.armada_client = armada_client
+        self.job_service = job_service_client
+        self.queue = queue
+        self.job_set_id = job_set_id
+        self.job_request_items = job_request_items
+
+    def execute(self, context):
+        """
+        Executes the Armada Operator.
+
+        Runs an Armada job and calls the job_service_client for polling.
+
+        """
+        job = self.armada_client.submit_jobs(
+            queue=self.queue,
+            job_set_id=self.job_set_id,
+            job_request_items=self.job_request_items,
+        )
+
+        try:
+            job_id = job.job_response_items[0].job_id
+            armada_logger.info(f"Running Armada job {self.name} with id {job_id}")
+        except Exception:
+            raise AirflowException("Armada has issues submitting job")
+        job_state, job_message = search_for_job_complete(
+            job_service_client=self.job_service,
+            queue=self.queue,
+            job_set_id=self.job_set_id,
+            airflow_task_name=self.name,
+            job_id=job_id,
+        )
+        armada_logger.info(
+            f"Armada Job finished with {job_state} and message: {job_message}"
```

Comment:
Interesting!  Thank you for that.  So fstrings execute but interpolation doesn't?

Reply:
I'll note this depends on the logging library, but in any I've looked at in python, it checks the log level first and will not interpolate the fields unless it's actually going to log it. This can (sadly) have a real performance difference on python -- it probably doesn't matter for the client, but it's good to know and a good practice to get into (using the log libraries' built-in interpolation.

If you're just using an f-string, it's going to have the fields inserted before the library is even called.